### ISSUE TITLE
Scale rounding issue in wlr-output-management-unstable-v1.xml

### ISSUE THREAD

**Issue by jameswalmsley** [swaywm/sway#4896]
See - #4877 - RE Kanshi scale config.

There is a mis-match between scale's set using swaymsg and e.g. kanshi which
uses the wlr-output-management protocol.

Kanshi parses the scale factor and converts to a float using strtof().

It is then converted by wayland to a fixed type (24 bits for integer, 8 bits for decimal precision).

wl_roots then deserializes this back to a float.

In my case 1.333333 is received by wl_roots as 1.332031.

2560 * (1.0/1.332031) == 1921.87719 ..
2560 * (1.0/1.333333) == 1920.00048 ..

I'd be happy to make PR to fix this.

What would be an acceptable way to fix this.
Send the scale as a string?

Or bundle the 32-bit float into a wayland int32 type?

Been using sway as my daily driver for over a year now. Absolutely love it,
and want to contribute.

**Comment by emersion** [comment]
Why is this such a big deal that the floating number doesn't have more precision?

**Comment by jameswalmsley** [comment]
How should I set my scale to 1.333 if the information is lost in the protocol?

I want the resultant rect to be 1920 not 1921.

From a user perspective setting 1.333 in kanshi should result in the same settings
as swaymsg.

Not sure what your question is, should I just accept that kanshi makes the scale
1.332 and not care what the rect is? Or are you saying that wl_roots/sway should
some how rectify this and give me the same result?

The float has enough precision, but the wayland protocol fixed point decimal does not for this case.

I think I started looking into this because some situations I had a 1 pixel space
between e.g. waybar and my application.
